packages/grid/veilid/server/veilid_streamer.py

```diff
@@ -0,0 +1,480 @@
+            total_chunks_count = self._calculate_chunks_count(message_size)
+
+            # Send STREAM_START request
+            stream_start_request = self.stream_start_struct.pack(
+                RequestType.STREAM_START,
+                call_id,
+                message_hash,
+                total_chunks_count,
+            )
+            await self._send_request(router, vld_key, stream_start_request)
+
+            # Send chunks
+            tasks = []
+            for chunk_number in range(total_chunks_count):
+                chunk = self._get_chunk(call_id, chunk_number, message)
+                tasks.append(self._send_request(router, vld_key, chunk))
+            await asyncio.gather(*tasks)
+
+            # Send STREAM_END request
+            stream_end_message = self.stream_end_struct.pack(
+                RequestType.STREAM_END, call_id
+            )
+            await self._send_request(router, vld_key, stream_end_message)
+
+        if is_request_stream:
+            # This is a new request stream, so we need to wait for
+            # the reply from the receiver
+            logger.debug("Waiting for reply...")
+            response = await buffer_for_holding_reply.message
+            logger.debug("Reply received")
+            # All operations finished, clean up the buffer and return the response
+            del self.buffers[call_id]
+            return response
+        return ResponseType.OK
+
+    async def receive_stream(
+        self,
+        connection: veilid.VeilidAPI,
+        router: veilid.RoutingContext,
+        update: veilid.VeilidUpdate,
+        callback: AsyncReceiveStreamCallback,
+    ) -> None:
+        """Receives a streamed message."""
+        message = update.detail.message
+        prefix = message[:8]
+
+        if prefix == RequestType.STREAM_SINGLE:
+            await self._handle_receive_stream_single(
+                connection, router, update, callback
+            )
+        elif prefix == RequestType.STREAM_START:
+            await self._handle_receive_stream_start(connection, update)
+        elif prefix == RequestType.STREAM_CHUNK:
+            await self._handle_receive_stream_chunk(connection, update)
+        elif prefix == RequestType.STREAM_END:
+            await self._handle_receive_stream_end(connection, router, update, callback)
+        else:
+            logger.error(f"[Bad Message] Message with unknown prefix: {prefix}")
+
+    @retry(veilid.VeilidAPIError, tries=4, delay=1, backoff=2)
+    async def _send_request(
+        self, router: veilid.RoutingContext, vld_key: str, request_data: bytes
+    ) -> None:
+        """Send an app call to the Veilid server and return the response."""
+        async with self._send_request_semaphore:
+            response = await router.app_call(vld_key, request_data)
+            if response != ResponseType.OK:
+                raise Exception("Unexpected response from server")
+
+    async def _send_response(
```

Comment:
MAYBE: we could also add retry logic to responses?

Reply:
Retrying a failed `app_call_reply` makes Veilid throw some operation id mismatch errors. We need to retry the original request in that case.

Reply:
Ahh ok 👍 .We could file a bug with veilid team on the same later?

Reply:
Yep 💯